### PR TITLE
Fixed Issue #151

### LINKED ISSUE #151: Use the named routes to navigate in between pages.
Use the named routes to navigate in between pages.

# Discussion 🗣
Use the named routes in constants.dart to navigate in between screen.

## Context
Make sure to use Replacement and RemoveUntil methods appropriately.

### PR DESCRIPTION
**Pull request by SapinderSingh**
### Related Issue
- Fixes #151 .

### Proposed Changes
- Added Named Routes
- Edited and Added some logic for routes and page controller.
- Removed some code from widget_test.dart because it was producing build errors because I passed arguments through named routes.
- Redundant Imports and code removed.

### Additional Information
- Some files and code were auto formatted by IDE. 

### Checklist
-  Tested
- No Conflicts
- Change In Code

### CODE REVIEW

File: lib/helpers/slide_route.dart [hunk 1]
```diff
@@ -102,4 +104,5 @@ class SlideBottomRoute extends PageRouteBuilder<SlideBottomRoute> {
           ),
         );
   final Widget page;
+  final RouteSettings settings;
```
Comment (himanshusharma89) [inline lib/helpers/slide_route.dart:107]: As mentioned previously don't create this settings property in this class, you can directly pass the object from route_page.dart

File: lib/main.dart [hunk 2]
```diff
@@ -50,10 +50,7 @@ class MyApp extends StatelessWidget {
             visualDensity: VisualDensity.adaptivePlatformDensity,
             textTheme:
                 GoogleFonts.poppinsTextTheme(Theme.of(context).textTheme)),
-        routes: <String, WidgetBuilder>{
-          '/login': (BuildContext context) => LoginScreen(),
-        },
-        home: LoginScreen(),
+        onGenerateRoute: RoutePage.generateRoute,
```
Comment (himanshusharma89) [inline lib/main.dart:53]: You need to add initialRoute here

File: lib/views/auth/login_view.dart [hunk 3]
```diff
@@ -242,12 +237,8 @@ class LoginScreenState extends State<LoginScreen> {
...
                               onTap: () {
                                 // ignore: always_specify_types
-                                Navigator.pushReplacement(context,
-                                    MaterialPageRoute<dynamic>(
-                                        builder: (BuildContext context) {
-                                  debugPrint('Navigate to sign up page');
-                                  return SignUpScreen();
-                                }));
+                                Navigator.of(context)
+                                    .pushNamed(RouteConstant.SIGN_UP_SCREEN);
```
Comment (himanshusharma89) [inline lib/views/auth/login_view.dart:241]: Use pushReplacementNamed here

File: lib/views/auth/login_view.dart [hunk 4]
```diff
@@ -76,37 +75,38 @@ class LoginScreenState extends State<LoginScreen> {
...
-                                          labelText: 'Email Address',
-                                          labelStyle: TextStyle(
-                                              fontSize: 20.0,
-                                              color: Colors.black),
-                                          filled: true,
-                                          fillColor: Colors.white,
-                                          border: OutlineInputBorder(
-                                              borderRadius:
-                                                  BorderRadius.zero))),
+                            ),
```
Comment (himanshusharma89) [inline lib/views/auth/login_view.dart:87]: What changes were required here? I don't see any fixes. If you want to formate the code use CTRL + F

File: lib/views/profile/settings/address_screen.dart [hunk 5]
```diff
@@ -22,11 +20,10 @@ class _AddressScreenState extends State<AddressScreen> {
           padding: const EdgeInsets.all(8.0),
           child: GestureDetector(
             onTap: () {
-              Navigator.pushAndRemoveUntil(context, MaterialPageRoute<dynamic>(
-                builder: (BuildContext context) {
-                  return Settings();
-                },
-              ), (Route<dynamic> route) => false);
+              Navigator.of(context).pushNamedAndRemoveUntil(
```
Comment (himanshusharma89) [inline lib/views/profile/settings/address_screen.dart:23]: Use Navigator.pop here

File: lib/views/profile/settings/manage_address.dart [hunk 6]
```diff
@@ -353,12 +353,10 @@ class _ManageAddressState extends State<ManageAddress> {
                           });
                         }
                         address = finalAddress;
-                        Navigator.pushAndRemoveUntil(context,
-                            MaterialPageRoute<dynamic>(
-                          builder: (BuildContext context) {
-                            return AddressScreen();
-                          },
-                        ), (Route<dynamic> route) => false);
+                        Navigator.of(context).pushNamedAndRemoveUntil(
```
Comment (himanshusharma89) [inline lib/views/profile/settings/manage_address.dart:356]: Use Navigator.pop here

File: lib/widgets/product/product_page.dart [hunk 7]
```diff
@@ -3,26 +3,12 @@ import 'package:google_mobile_ads/google_mobile_ads.dart';
...
 import 'package:retro_shopping/helpers/ad_state.dart';
 import 'package:retro_shopping/helpers/constants.dart';
-
+import 'package:retro_shopping/model/product_model.dart';
 import '../../helpers/app_icons.dart';
 import '../retro_button.dart';
 
 // ignore: must_be_immutable
 class ProductPage extends StatefulWidget {
-  ProductPage(
```
Comment (himanshusharma89) [inline lib/widgets/product/product_page.dart:12]: You can have the Product Object here, which will get the object from the route_page.dart

File: lib/widgets/product/product_page.dart [hunk 8]
```diff
@@ -50,6 +36,7 @@ class _ProductPageState extends State<ProductPage> {
 
   @override
   Widget build(BuildContext context) {
+    final Product args = ModalRoute.of(context).settings.arguments as Product;
```
Comment (himanshusharma89) [inline lib/widgets/product/product_page.dart:39]: As mentioned use the Product Object instead of this.

**Review by himanshusharma89 — CHANGES_REQUESTED** [review]

**Comment by himanshusharma89** [comment]
@SapinderSingh Resolve the updates and don't close this PR

File: lib/helpers/route_page.dart [hunk 9]
```diff
@@ -52,33 +62,50 @@ class RoutePage {
         );
       case RouteConstant.PRODUCTS_SCREEN:
         return SlideBottomRoute(
+          settings: settings,
```
[the remark this answers is not in the document]
Comment (SapinderSingh) [reply lib/helpers/route_page.dart:65]: @himanshusharma89 I didn't understand. Please help

[the remark this answers is not in the document]
Comment (himanshusharma89) [reply lib/helpers/route_page.dart:65]: Before the return statement, you can create a Product object which will have the value from the settings. Check this: https://github.com/himanshusharma89/Flutter-Blog-App/blob/d872f9974622c7072c77fe3a6892ebe290cda1ba/lib/helpers/route_page.dart#L33

[the remark this answers is not in the document]
Comment (SapinderSingh) [reply lib/helpers/route_page.dart:65]: Thank You. I got it.

File: lib/views/auth/login_view.dart (hunk 3 above)
Comment (SapinderSingh) [reply lib/views/auth/login_view.dart:241]: Well, I think we should not close the app if user presses back button on the sign up page.

Comment (SapinderSingh) [reply lib/views/auth/login_view.dart:241]: because of don't have a "login instead" button on the sign up screen

Comment (himanshusharma89) [reply lib/views/auth/login_view.dart:241]: Okay, then let it be, I thought we have a toggle view in between the screens.

**Review by SapinderSingh — COMMENTED** [review]
I have made the changes. Please review the PR

**Pull request merged**